src/validate.rs

```diff
@@ -10,14 +10,361 @@ use crate::base::name::Name;
 use crate::base::name::ToName;
 use crate::base::rdata::{ComposeRecordData, RecordData};
 use crate::base::record::Record;
+use crate::base::scan::IterScanner;
 use crate::base::wire::{Compose, Composer};
 use crate::rdata::{Dnskey, Rrsig};
 use bytes::Bytes;
 use octseq::builder::with_infallible;
 use ring::{digest, signature};
+use std::boxed::Box;
 use std::vec::Vec;
 use std::{error, fmt};
 
+/// A generic public key.
+#[derive(Clone, Debug)]
+pub enum PublicKey {
```

Comment:
I'm not sure about PublicKey being in the validate module. For example I want to construct a key for signing and for that I need the public part as well as the private part, and to use validate for that seems strange. Could this not go in src/sign/key.rs?

Reply:
The DNSKEY record conversion that `PublicKey` offers is also relevant to validation.  Furthermore, I want to put most of the NSEC3 signing logic (everything up to `Sign::sign()`) in `validate` since they'll use it for validation too.